nemo/collections/asr/models/asr_model.py

```diff
@@ -136,9 +136,56 @@ def on_after_backward(self):
                 torch.distributed.all_reduce(valid_gradients, op=torch.distributed.ReduceOp.MIN)
 
             if valid_gradients < 1:
-                logging.warning(f'detected inf or nan values in gradients! Setting gradients to zero.')
+                logging.warning('detected inf or nan values in gradients! Setting gradients to zero.')
                 self.zero_grad()
 
+    def on_train_start(self):
+        super().on_train_start()
+
+        # dynamic freezing
+        # should fire only once, on the very first batch of training and never again
+        if not hasattr(self, '_freeze_cfg'):
```

Comment:
One suggestion I have is to drop the enabled from the config and have something like this where -1 or None means not any freezing. It can make the config simpler. It is your call as the current one should also be OK:
```
  freeze_modules:
     preprocessor: -1
     encoder: -1
     joint: -1
```

Reply:
I originally wanted to do this, but Som explicitly requested that I add that global enabled flag. @titu1994 Can you please explain why we need that global enabled flag and can't do it like Vahid suggests?

Reply:
Yould have to update N different flags to disable it then, vs one arg for enabled=False